File: php/src/Resource/Runs.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Resource;

use Fivexer\SDK\Fivexer;
use Fivexer\SDK\Model\WorkflowRun;
use Fivexer\SDK\Model\WorkflowRunPage;
use Fivexer\SDK\Model\WorkflowRunSteps;

/** Workflow runs across the workspace, including the external (callback) step outcomes. */
final class Runs
{
    public function __construct(private readonly Fivexer $client)
    {
    }

    public function list(
        ?string $status = null,
        ?string $workflowId = null,
        ?string $cursor = null,
        ?int $limit = null,
    ): WorkflowRunPage {
        $data = $this->client->request('GET', '/workflow-runs', null, [
            'status' => $status,
            'workflowId' => $workflowId,
            'cursor' => $cursor,
            'limit' => $limit,
        ]) ?? [];
        return WorkflowRunPage::fromArray($data);
    }

    public function get(string $runId): WorkflowRun
    {
        return WorkflowRun::fromArray(
            $this->client->request('GET', '/workflow-runs/' . \rawurlencode($runId)) ?? []
        );
    }

    /** The per-step state view — what a canvas UI paints. */
    public function steps(string $runId): WorkflowRunSteps
    {
        return WorkflowRunSteps::fromArray(
            $this->client->request('GET', '/workflow-runs/' . \rawurlencode($runId) . '/steps') ?? []
        );
    }

    public function cancel(string $runId): WorkflowRun
    {
        return WorkflowRun::fromArray(
            $this->client->request('POST', '/workflow-runs/' . \rawurlencode($runId) . '/cancel', []) ?? []
        );
    }

    /**
     * Complete an external (callback) step, advancing the run.
     *
     * @param array<string, mixed>|null $data
     */
    public function completeStep(string $runId, string $stepId, ?array $data = null): WorkflowRun
    {
        $body = $data === null ? [] : ['data' => $data];
        $path = '/workflow-runs/' . \rawurlencode($runId) . '/steps/' . \rawurlencode($stepId) . '/complete';
        return WorkflowRun::fromArray($this->client->request('POST', $path, $body) ?? []);
    }

    /** Fail an external (callback) step. */
    public function failStep(string $runId, string $stepId, ?string $error = null): WorkflowRun
    {
        $body = $error === null ? [] : ['error' => $error];
        $path = '/workflow-runs/' . \rawurlencode($runId) . '/steps/' . \rawurlencode($stepId) . '/fail';
        return WorkflowRun::fromArray($this->client->request('POST', $path, $body) ?? []);
    }
}

File: java/src/main/java/io/fivexer/sdk/Runs.java
package io.fivexer.sdk;

import io.fivexer.sdk.internal.Json;
import io.fivexer.sdk.model.WorkflowRun;
import io.fivexer.sdk.model.WorkflowRunPage;
import io.fivexer.sdk.model.WorkflowRunSteps;
import java.util.Map;

/** Workflow runs across the workspace, including the external (callback) step outcomes. */
public final class Runs {

    private final Fivexer client;

    Runs(Fivexer client) {
        this.client = client;
    }

    public WorkflowRunPage list(ListRunsQuery query) {
        ListRunsQuery q = query == null ? new ListRunsQuery() : query;
        return client.request("GET", "/workflow-runs", null, q.toQuery(), WorkflowRunPage.class, false);
    }

    public WorkflowRunPage list() {
        return list(null);
    }

    public WorkflowRun get(String runId) {
        return client.request("GET", "/workflow-runs/" + Json.enc(runId), null, null,
                WorkflowRun.class, false);
    }

    /** The per-step state view — what a canvas UI paints. */
    public WorkflowRunSteps steps(String runId) {
        return client.request("GET", "/workflow-runs/" + Json.enc(runId) + "/steps", null, null,
                WorkflowRunSteps.class, false);
    }

    public WorkflowRun cancel(String runId) {
        return client.request("POST", "/workflow-runs/" + Json.enc(runId) + "/cancel", null, null,
                WorkflowRun.class, false);
    }

    /** Complete an external (callback) step, advancing the run. */
    public WorkflowRun completeStep(String runId, String stepId, Map<String, Object> data) {
        String body = data == null ? "{}" : Json.object("data", data);
        return client.request("POST",
                "/workflow-runs/" + Json.enc(runId) + "/steps/" + Json.enc(stepId) + "/complete",
                body, null, WorkflowRun.class, false);
    }

    public WorkflowRun completeStep(String runId, String stepId) {
        return completeStep(runId, stepId, null);
    }

    /** Fail an external (callback) step. */
    public WorkflowRun failStep(String runId, String stepId, String error) {
        String body = error == null ? "{}" : Json.object("error", error);
        return client.request("POST",
                "/workflow-runs/" + Json.enc(runId) + "/steps/" + Json.enc(stepId) + "/fail",
                body, null, WorkflowRun.class, false);
    }

    public WorkflowRun failStep(String runId, String stepId) {
        return failStep(runId, stepId, null);
    }
}
